src/kernels/MIOpenAdam.cpp

```diff
@@ -0,0 +1,226 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2023 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+
+template <typename T1, typename T2>
+inline __device__ void AdamInternal(T1* param_in,
+                                    T1* param_out,
+                                    T1* exp_avg_in,
+                                    T1* exp_avg_out,
+                                    T1* exp_avg_sq_in,
+                                    T1* exp_avg_sq_out,
+                                    T1* max_exp_avg_sq_in,
+                                    T1* max_exp_avg_sq_out,
+                                    T2 grad,
+                                    T2 lr,
+                                    T2 beta1,
+                                    T2 beta2,
+                                    T2 weight_decay,
+                                    T2 eps,
+                                    int step,
+                                    bool amsgrad,
+                                    bool maximize,
+                                    size_t gid)
+{
+    T2 param      = static_cast<T2>(param_in[gid]);
+    T2 exp_avg    = static_cast<T2>(exp_avg_in[gid]);
+    T2 exp_avg_sq = static_cast<T2>(exp_avg_sq_in[gid]);
+
+    __builtin_assume(exp_avg_sq >= 0 && exp_avg_sq <= 1);
+    __builtin_assume(step >= 0);
```

Comment:
Is the `step` should be an `int`? In case of `unsigned int` that assumption is automatically true.

Reply:
fixed it, thank you.